File: packages/py/analytics/src/bolsa_analytics/cognitive/evidence_engine.py
"""Evidence Engine v1 — suite estadística + ART-EDGE-REPORT (RFC-008 D3)."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any

import numpy as np

from bolsa_analytics.cognitive.edge_report import (
    EdgeReport,
    StatisticalSuiteResult,
    build_edge_report,
)
from bolsa_analytics.cognitive.psr_dsr import psr_dsr_from_returns
from bolsa_analytics.cognitive.stats_suite import (
    monte_carlo_permutation_p_value,
    walk_forward_efficiency,
)
from bolsa_analytics.cognitive.trials_log import TrialsLog


@dataclass(frozen=True, slots=True)
class EvidenceEngineInput:
    strategy_or_signal_ref: str
    trade_returns: Sequence[float]
    trials_log: TrialsLog
    in_sample_sharpe: float | None = None
    out_of_sample_sharpe: float | None = None
    # Lab optimize WFE (score OOS/IS). Preferred over Sharpe WFE when set.
    lab_walk_forward_efficiency: float | None = None
    stress_survival_rate: float | None = None
    bootstrap_alpha_ci_lower: float | None = None
    bootstrap_alpha_ci_upper: float | None = None
    monte_carlo_permutations: int = 1000
    seed: int = 42


@dataclass(frozen=True, slots=True)
class EvidenceEngineResult:
    edge_report: EdgeReport
    trials_n: int
    trials_log_id: str
    auto_live_eligible: bool
    block_reasons: tuple[str, ...]

    def to_dict(self) -> dict[str, Any]:
        return {
            "edgeReport": self.edge_report.to_dict(),
            "trialsN": self.trials_n,
            "trialsLogId": self.trials_log_id,
            "autoLiveEligible": self.auto_live_eligible,
            "blockReasons": list(self.block_reasons),
        }


def _win_rate(returns: Sequence[float]) -> float | None:
    arr = np.asarray(list(returns), dtype=float)
    if arr.size == 0:
        return None
    return float(np.mean(arr > 0))


def run_evidence_suite(inp: EvidenceEngineInput) -> EvidenceEngineResult:
    """
    Ejecuta WFE + Monte Carlo + PSR/DSR (+ opcionales) y emite EdgeReport.
    Sin trials_n >= 1 el DSR no es válido → auto_live_eligible = False.
    """
    trials_n = inp.trials_log.trials_n
    returns = list(inp.trade_returns)
    notes: list[str] = []

    if trials_n < 1:
        notes.append("TrialsLog vacío — DSR inválido; auto-live bloqueado")
        suite = StatisticalSuiteResult(
            trials_n=0,
            sample_trades_count=len(returns),
            historical_win_rate=_win_rate(returns),
        )
        report = build_edge_report(inp.strategy_or_signal_ref, suite, notes=tuple(notes))
        return EvidenceEngineResult(
            edge_report=report,
            trials_n=0,
            trials_log_id=inp.trials_log.log_id,
            auto_live_eligible=False,
            block_reasons=("missing_trials_log",),
        )

    wfe = None
    wfe_source = None
    if inp.lab_walk_forward_efficiency is not None:
        wfe = float(inp.lab_walk_forward_efficiency)
        wfe_source = "lab_score"
        notes.append(
            "WFE from optimize lab (score OOS/IS ratio), not Sharpe_OOS/Sharpe_IS"
        )
    elif inp.in_sample_sharpe is not None and inp.out_of_sample_sharpe is not None:
        wfe = walk_forward_efficiency(inp.in_sample_sharpe, inp.out_of_sample_sharpe)
        wfe_source = "sharpe"

    mc_p = (
        monte_carlo_permutation_p_value(
            returns,
            permutations=inp.monte_carlo_permutations,
            seed=inp.seed,
        )
        if len(returns) >= 3
        else 1.0
    )

    _sr, psr, dsr = psr_dsr_from_returns(returns, trials_n=trials_n)

    suite = StatisticalSuiteResult(
        trials_n=trials_n,
        walk_forward_efficiency=wfe,
        wfe_source=wfe_source,
        monte_carlo_p_value=mc_p,
        psr=round(psr, 4),
        dsr=round(dsr, 4),
        bootstrap_alpha_ci_lower=inp.bootstrap_alpha_ci_lower,
        bootstrap_alpha_ci_upper=inp.bootstrap_alpha_ci_upper,
        stress_survival_rate=inp.stress_survival_rate,
        historical_win_rate=_win_rate(returns),
        sample_trades_count=len(returns),
    )
    report = build_edge_report(inp.strategy_or_signal_ref, suite, notes=tuple(notes))

    block: list[str] = []
    if report.band == "luck":
        block.append("edge_band_luck")
    if report.credibility < 65:
        block.append("credibility_below_65")
    if dsr < 0.5:
        block.append("dsr_below_0.5")
    if mc_p > 0.05:
        block.append("monte_carlo_p_above_0.05")
    if wfe is not None and wfe < 0.5:
        block.append("wfe_below_0.5")

    eligible = len(block) == 0 and report.band in {"skill", "uncertain"}
    # uncertain may still be paper-only; auto-live requires skill by default
    if report.band != "skill":
        block.append("band_not_skill_for_auto_live")
        eligible = False

    return EvidenceEngineResult(
        edge_report=report,
        trials_n=trials_n,
        trials_log_id=inp.trials_log.log_id,
        auto_live_eligible=eligible,
        block_reasons=tuple(block),
    )

File: packages/shared/src/cognitive/evidence-engine.ts
/**
 * Evidence Engine types + ART-EDGE-REPORT (RFC-008 D3 skeleton).
 * Responde: ¿cuánto puedo creerme esta señal? (Credibility), no solo “¿funciona?”.
 */

export type EvidenceDirection = 'supports' | 'contradicts' | 'neutral';

export type EvidenceKind =
  | 'data_quality'
  | 'market_regime'
  | 'technical'
  | 'fundamental'
  | 'opportunity'
  | 'news_event'
  | 'macro'
  | 'risk'
  | 'policy'
  | 'statistical';

export interface EvidenceV1 {
  evidenceId: string;
  evidenceKind: EvidenceKind;
  claim: string;
  direction: EvidenceDirection;
  weight: number;
  confidence: number;
  validFrom?: string;
  validTo?: string;
  /** Half-life in hours */
  decayHalfLifeHours?: number;
  refs?: Record<string, string>;
}

export interface EvidenceBundleV1 {
  artifactType: 'ART-EVIDENCE-BUNDLE';
  schemaVersion: '1.0.0';
  bundleId: string;
  instrumentId: string;
  timestamp: string;
  evidences: EvidenceV1[];
}

export type EdgeBand = 'skill' | 'uncertain' | 'luck';

export type WfeSource = 'lab_score' | 'sharpe';

export interface StatisticalSuiteResult {
  /** Walk-Forward Efficiency — Sharpe ratio or lab score ratio (see wfeSource). */
  walkForwardEfficiency?: number | null;
  /** Provenance of walkForwardEfficiency. */
  wfeSource?: WfeSource | null;
  monteCarloPValue?: number | null;
  /** Probabilistic Sharpe Ratio */
  psr?: number | null;
  /** Deflated Sharpe Ratio — inválido sin trialsN */
  dsr?: number | null;
  bootstrapAlphaCiLower?: number | null;
  bootstrapAlphaCiUpper?: number | null;
  stressSurvivalRate?: number | null;
  historicalWinRate?: number | null;
  sampleTradesCount?: number | null;
  /** Nº de trials / hipótesis exploradas (obligatorio para DSR) */
  trialsN: number;
}

export interface EdgeReportV1 {
  artifactType: 'ART-EDGE-REPORT';
  schemaVersion: '1.0.0';
  edgeReportId: string;
  version: string;
  strategyOrSignalRef: string;
  instrumentUniverseRef?: string;
  createdAt: string;
  suite: StatisticalSuiteResult;
  /** Credibility agregada 0–100 */
  credibility: number;
  edgeScore: number;
  band: EdgeBand;
  notes?: string[];
}

export interface CredibilityWeights {
  wMonteCarlo: number;
  wWfe: number;
  wDsr: number;
  wBootstrap: number;
  wStress: number;
}

export const DEFAULT_CREDIBILITY_WEIGHTS: CredibilityWeights = {
  wMonteCarlo: 0.25,
  wWfe: 0.25,
  wDsr: 0.2,
  wBootstrap: 0.15,
  wStress: 0.15,
};

function clamp01(n: number): number {
  return Math.min(1, Math.max(0, n));
}

/**
 * Credibility 0–100 a partir de la suite.
 * DSR se ignora (contribuye 0) si trialsN < 1 — el caller debe registrar trials.
 */
export function computeCredibility(
  suite: StatisticalSuiteResult,
  weights: CredibilityWeights = DEFAULT_CREDIBILITY_WEIGHTS,
): { credibility: number; edgeScore: number; band: EdgeBand } {
  const mc =
    suite.monteCarloPValue == null
      ? 0
      : clamp01(1 - suite.monteCarloPValue / 0.05) * (suite.monteCarloPValue <= 0.05 ? 1 : 0.3);

  const wfe =
    suite.walkForwardEfficiency == null
      ? 0
      : clamp01(suite.walkForwardEfficiency);

  const dsr =
    suite.trialsN < 1 || suite.dsr == null ? 0 : clamp01(suite.dsr);

  let bootstrap = 0;
  if (
    suite.bootstrapAlphaCiLower != null &&
    suite.bootstrapAlphaCiUpper != null
  ) {
    bootstrap = suite.bootstrapAlphaCiLower > 0 ? 1 : 0.35;
  }

  const stress =
    suite.stressSurvivalRate == null ? 0 : clamp01(suite.stressSurvivalRate);

  const sumW =
    weights.wMonteCarlo +
    weights.wWfe +
    weights.wDsr +
    weights.wBootstrap +
    weights.wStress;

  const score01 =
    (weights.wMonteCarlo * mc +
      weights.wWfe * wfe +
      weights.wDsr * dsr +
      weights.wBootstrap * bootstrap +
      weights.wStress * stress) /
    sumW;

  const credibility = Math.round(score01 * 1000) / 10;
  const edgeScore = credibility;
  const band: EdgeBand =
    credibility >= 85 ? 'skill' : credibility >= 65 ? 'uncertain' : 'luck';

  return { credibility, edgeScore, band };
}

export interface TrialRecordV1 {
  trialId: string;
  hypothesisRef: string;
  paramsHash: string;
  sharpeIs?: number | null;
  createdAt: string;
  notes?: string | null;
}

/** Log de hipótesis — N = trials.length; obligatorio para DSR. */
export interface TrialsLogV1 {
  logId: string;
  strategyFamilyRef: string;
  trialsN: number;
  trials: TrialRecordV1[];
}

export interface AutoLiveCheckV1 {
  allowed: boolean;
  reasons: string[];
  policyId: string;
  edgeReportId?: string | null;
  credibility?: number | null;
}

/** Suite D3 — implementación numérica en Python (`bolsa_analytics.cognitive`). */
export interface EvidenceEngineSuiteApi {
  monteCarloPermutationPValue(tradeReturns: number[], permutations?: number): number;
  walkForwardEfficiency(inSampleSharpe: number, outOfSampleSharpe: number): number;
  /** Requiere trialsN > 0 para DSR válido */
  buildEdgeReport(input: {
    strategyOrSignalRef: string;
    suite: StatisticalSuiteResult;
  }): EdgeReportV1;
}
